catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/monitor/impl/WorkspaceQueryService.java

```diff
@@ -0,0 +1,386 @@
+                        .getId(), new ImmutablePair<>(pair.getLeft(), pair.getRight())));
+
+        return results;
+    }
+
+    private Pair<WorkspaceMetacardImpl, Long> getTaskResult(
+            ForkJoinTask<Pair<WorkspaceMetacardImpl, Long>> workspaceTask, long timeout,
+            TimeUnit timeoutUnit) {
+        try {
+            return workspaceTask.get(timeout, timeoutUnit);
+        } catch (TimeoutException e) {
+            LOGGER.warn("Timeout", e);
+        } catch (ExecutionException | InterruptedException e) {
+            LOGGER.warn("ForkJoinPool error", e);
+        }
+        return null;
+    }
+
+    private List<WorkspaceTask> createWorkspaceTasks(
+            Map<String, Pair<WorkspaceMetacardImpl, List<QueryMetacardImpl>>> queryMetacards) {
+
+        List<WorkspaceTask> workspaceTasks = new ArrayList<>();
+
+        for (Map.Entry<String, Pair<WorkspaceMetacardImpl, List<QueryMetacardImpl>>> workspaceQueryEntry : queryMetacards.entrySet()) {
+            Map<String, List<QueryMetacardImpl>> queryMetacardsGroupedBySource = groupBySource(
+                    workspaceQueryEntry.getValue()
+                            .getRight());
+            List<QueryRequest> queryRequests =
+                    getQueryRequests(queryMetacardsGroupedBySource.values()
+                            .stream());
+            if (!queryRequests.isEmpty()) {
+                workspaceTasks.add(new WorkspaceTask(workspaceQueryEntry.getValue()
+                        .getLeft(), queryRequests));
+            }
+        }
+
+        return workspaceTasks;
+    }
+
+    private Map<String, List<QueryMetacardImpl>> groupBySource(
+            List<QueryMetacardImpl> queryMetacards) {
+        final Map<String, List<QueryMetacardImpl>> groupedBySource = new HashMap<>();
+        for (QueryMetacardImpl queryMetacard : queryMetacards) {
+
+            List<String> sources = queryMetacard.getSources();
+
+            if (!sources.isEmpty()) {
+                sources.forEach(sourceId -> groupedBySource.compute(sourceId,
+                        addToList(queryMetacard)));
+            } else {
+                groupedBySource.compute(UNKNOWN_SOURCE, addToList(queryMetacard));
+            }
+        }
+        return groupedBySource;
+    }
+
+    private BiFunction<String, List<QueryMetacardImpl>, List<QueryMetacardImpl>> addToList(
+            QueryMetacardImpl queryMetacard) {
+        return (id, queries) -> {
+            if (queries == null) {
+                return Lists.newArrayList(queryMetacard);
+            } else {
+                queries.add(queryMetacard);
+                return queries;
+            }
+        };
+    }
+
+    private List<QueryRequest> getQueryRequests(
+            Stream<List<QueryMetacardImpl>> queriesGroupedBySource) {
+
+        final Filter modifiedFilter = filterService.getModifiedDateFilter(getOneDayBack());
+
+        return queriesGroupedBySource.map(queriesForSource -> queriesForSource.stream()
```

Comment:
Might want to extract the lambda inside this `map()` call into its own function.